ast/src/analyzed/mod.rs

```diff
@@ -1498,28 +1505,45 @@ impl<T> AlgebraicExpression<T> {
     }
 
     /// Returns the degree of the expressions
-    pub fn degree(&self) -> usize {
+    pub fn degree_with_cache(
+        &self,
+        intermediate_definitions: &BTreeMap<AlgebraicReferenceThin, AlgebraicExpression<T>>,
+        cache: &mut BTreeMap<AlgebraicReferenceThin, usize>,
+    ) -> usize {
         match self {
-            AlgebraicExpression::Reference(reference) => {
-                // We don't have access to the definitions of intermediate polynomials here, so we can't know their degree.
-                assert!(
-                    matches!(
-                        reference.poly_id.ptype,
-                        PolynomialType::Committed | PolynomialType::Constant
-                    ),
-                    "Intermediate polynomials should have been inlined."
-                );
-                // Fixed and witness columns have degree 1
-                1
-            }
+            AlgebraicExpression::Reference(reference) => match reference.poly_id.ptype {
+                PolynomialType::Committed | PolynomialType::Constant => 1,
+                PolynomialType::Intermediate => {
+                    let reference = reference.to_thin();
+                    let value = cache.get(&reference).cloned();
+                    match value {
+                        Some(v) => v,
+                        None => {
+                            let def = intermediate_definitions
+                                .get(&reference)
+                                .expect("Intermediate definition not found.");
+                            let result = def.degree_with_cache(intermediate_definitions, cache);
+                            cache.insert(reference, result);
+                            result
+                        }
+                    }
```

Comment:
Does this work with `unwrap_or_else` or is there an issue with the borrow checker?

Reply:
Nope, works!